### PR TITLE
new cltv guide

### PR DESCRIPTION
**Pull request by bucko13**
A guide for creating time locked transactions. Includes guide for mock transactions and live transactions on a regtest network. Would love any feedback on where this can be improved in terms of code or clarity of explanation!

### CODE REVIEW

**Comment by bucko13** [comment]
This solves #45

File: guides-markdown/cltv.md [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+const keyring2 = KeyRing.generate(true);
+// can only be redeemed after the 100th block has been mined
+const locktime = '100';
+keyring.witness = true;
+keyring2.witness = true;
+
+// 2) Get hash and save it to keyring
+const pkh = keyring.getKeyHash();
+const script = createScript(locktime, pkh);
+script.compile();
```
Comment (pinheadmz) [inline guides-markdown/cltv.md]: `script` is already compiled in `createScript()` L#141 do you need it again here?

File: guides-markdown/cltv.md [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+
+  const redeem = input.witness.getRedeem();
+
+  assert(
+    redeem,
+    'Witness has not been templated'
+  );
+
+  witness = input.witness;
+  version = 1;
```
Comment (pinheadmz) [inline guides-markdown/cltv.md]: maybe indicate that this version number is for sighash type (segwit vs legacy -- and not "witness program version") I had to look it up :-)

File: guides-markdown/cltv.md [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+      console.log('transaction sent to mempool');
+
+      // save the transaction information to to a file
+      txInfo = { lockedTx, lockingAddr, redeemScript, locktime, redeemAddress: address };
+      fs.writeFileSync(txInfoPath, JSON.stringify(txInfo, null, 2));
+
+      // mine one block to get tx on chain
+      // make sure you're doing this on regtest or simnet and not testnet or mainnet
+      // you can also use bPanel and the @bpanel/simple-mining plugin to do this instead
+      const minedBlock = await nodeClient.execute('generate', [1]);
```
Comment (pinheadmz) [inline guides-markdown/cltv.md]: dunno if you need to mention that `rpc generate` will error if `--coinbase-address` isn't set at launch.

**Comment by bucko13** [comment]
@pinheadmz thanks for the comments. Updated.

File: guides-markdown/cltv.md [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+cb.addInput({
+  prevout: new Outpoint(),
+  script: new Script(),
+  sequence: 0xffffffff
+});
+
+// Send 50,000 satoshis to our locking address.
+// this will lock up the funds to whoever can solve
+// the CLTV script
+cb.addOutput(lockingAddr, lockingValue);
```
Comment (pinheadmz) [inline guides-markdown/cltv.md]: `lockingValue` is undefined

File: guides-markdown/cltv.md [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+// You can test if the CLTV script is working or not
+// by changing this to a value less than what our script requires
+// which will cause the `mtx.verify` call to fail below
+mtx.setLocktime(parseInt(locktime));
+
+// if you console log the input being signed,
+// you'll notice it now has a witness stack and redeem script
+// before script, witness, and redeem were empty
+console.log('signed mtx:', mtx);
+assert(mtx.verify(flags), 'mtx did not verify');
```
Comment (pinheadmz) [inline guides-markdown/cltv.md]: `flags` is undefined

File: guides-markdown/cltv.md [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+of a previous transaction, making that the input for a new transaction,
+and, then satisfying some locking condition on that previous output.
+"Scripts" are conditions that need to be satisfied on an output to
+prove ownership. You can have a Bitcoin script that is locked with the
+math problem "What is 2 + 5", and anyone that knows to answer "7" can "prove"
+ownership over that output (Check out this guide, [Intro To Scripting](http://bcoin.io/guides/scripting.html)
+to see how to write and redeem this simple script in bcoin, and [this section](https://github.com/bitcoinbook/bitcoinbook/blob/develop/ch06.asciidoc#a-simple-script) in Mastering Bitcoin for a helpful walkthrough).
+
+In the vast majority of Bitcoin and other cryptocurrency transactions,
+you prove ownership by signing a transaction input with the private key (basically a password that is stored in your wallet) that corresponds to the address that the source output was sent to. I.e. The UTXO is locked by the requirement that some signature
```
Comment (tynes) [inline guides-markdown/cltv.md]: Just a nit, this line is much longer than the other lines

File: guides-markdown/cltv.md [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+[See BIP-065 Specs](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki)).
+
+```javascript
+/**
+ * @param {String} locktime - Time that the script can not
+ * be redeemed before
+ * @param {Buffer} public key hash
+ * @returns {Script}
+**/
+function createScript(locktime='100', publicKeyHash) {
```
Comment (tynes) [inline guides-markdown/cltv.md]: Is it possible to use the default value of `locktime` in this case when its the first argument?

```
createScript(,'xxx')
```

I don't think that you should give it a default value and that you should `assert` that it was passed in

Comment (bucko13) [reply guides-markdown/cltv.md]: Good call!

**Comment by bucko13** [comment]
Merging this in. If there are any other comments let me know and I can post an update.

**Pull request merged**